### PR TITLE
[FLOC-1643] Disable SELINUX

### PR DESCRIPTION
**Pull request by wallrj**
Fixes https://clusterhq.atlassian.net/browse/FLOC-1643

This disables SELINUX for the current  environment and upon reboot, by modifying the /etc/selinux/config file.

I didn't want to force the user to reboot again so I added both commands.

### CODE REVIEW

File: docs/indepth/installation.rst [hunk 1]
```diff
@@ -374,6 +366,16 @@ Paste them into a root console on the target node:
 Post installation configuration for Fedora 20 and CentOS 7
 ----------------------------------------------------------
 
+First disable SELINUX.
```
Comment (exarkun) [inline docs/indepth/installation.rst]: I think `SELinux` is the preferred markup.

**Comment by wallrj** [comment]
Note that the selinux status will be set to "Permissive" until the server is rebooted, at which point it's disabled.

```
[root@ip-172-31-19-200 ~]# sestatus 
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             targeted
Current mode:                   permissive
Mode from config file:          disabled
Policy MLS status:              enabled
Policy deny_unknown status:     allowed
Max kernel policy version:      29
[root@ip-172-31-19-200 ~]# ^C
[root@ip-172-31-19-200 ~]# shutdown -r now
Write failed: Broken pipe
(1643)(release-maintenance/flocker-0.4.0pre1/disable-selinux-FLOC-1643 ? ▴2)[~/.../HybridLogic/flocker]$ ssh root@54.191.164.117
Last login: Thu Apr  2 16:02:04 2015 from 151.237.232.130
[root@ip-172-31-19-200 ~]# sestatus 
SELinux status:                 disabled

```

File: docs/indepth/installation.rst [hunk 2]
```diff
@@ -374,6 +366,16 @@ Paste them into a root console on the target node:
...
+
+.. task:: disable_selinux
+   :prompt: [root@node]#
+
+.. note:: Flocker does not currently set the necessary SELINUX context types on the filesystem mountpoints that it creates on nodes.
+          This prevents Docker containers accessing those filesystems as volumes.
+          A future stable version of Flocker will set the appropriate SELINUX contexts.
```
Comment (exarkun) [inline docs/indepth/installation.rst]: Ooh.  A forward-looking statement.  Quite bold of you.

I suggest making this less specific.  A statement like "A future version of Flocker may provide a different integration strategy." or something else weasel-y (on the basis that we don't actually know what we're going to do yet).

File: flocker/provision/_install.py [hunk 3]
```diff
@@ -212,6 +212,18 @@ def task_install_kernel_devel():
...
+    Disable SELINUX for this session and permanently.
+    XXX: Remove this when we work out suitable SELINUX settings.
+    See https://clusterhq.atlassian.net/browse/FLOC-619.
+    """
+    return [
+        Run(command="setenforce 0"),
+        Run(command="test -e /etc/selinux/config && sed --in-place='.preflocker' 's/^SELINUX=.*$/SELINUX=disabled/g' /etc/selinux/config"),
+    ]
+
+
```
Comment (exarkun) [inline flocker/provision/_install.py]: Despite this, I can't seem to build the docs:

```
reading sources... [ 76%] indepth/installation                                                                                                                 
Exception occurred:
  File "/home/jean-paul/Projects/flocker/develop/flocker/provision/_sphinx.py", line 66, in run
    task = getattr(tasks, 'task_%s' % (self.arguments[0],))
AttributeError: 'module' object has no attribute 'task_disable_selinux'
The full traceback has been saved in /tmp/sphinx-err-nY3BcH.log, if you want to report the issue to the developers.
```

**Comment by exarkun** [comment]
Thanks.  As a design, I'm happy to accept this.  It's almost an acceptable implementation, too.  There are some doc build issues that prevent me from seeing what the generated docs look like.  I'll review that when you re-submit for code review though.  Thanks.

File: flocker/provision/_install.py (hunk 3 above)
Comment (wallrj) [inline flocker/provision/_install.py]: I don't understand why the docs failed to build. Here's where the new task is imported, so it should be available for use in the `_sphinx.py` module. 
 * https://github.com/ClusterHQ/flocker/pull/1270/files#diff-6a6e7e9f8092167d5a5c2ec85ea37341R9

Documentation builds for me.

**Comment by wallrj** [comment]
Thanks @exarkun 

I addressed your comments and double checked that the documentation builds.
Perhaps you had some stale pyc files or perhaps you needed to re-install flocker in your virtualenv.

The buildbot built docs are here:
 * http://build.clusterhq.com/results/docs/release-maintenance/flocker-0.4.0pre1/disable-selinux-FLOC-1643/build-5484/indepth/installation.html#post-installation-configuration-for-fedora-20-and-centos-7

File: docs/indepth/installation.rst [hunk 4]
```diff
@@ -374,6 +366,16 @@ Paste them into a root console on the target node:
 Post installation configuration for Fedora 20 and CentOS 7
 ----------------------------------------------------------
 
+First disable SELinux.
+
+.. task:: disable_selinux
+   :prompt: [root@node]#
+
+.. note:: Flocker does not currently set the necessary SELinux context types on the filesystem mount points that it creates on nodes.
+          This prevents Docker containers accessing those filesystems as volumes.
```
Comment (exarkun) [inline docs/indepth/installation.rst]: `... containers from accessing ...` (I think omitting the "from" is a British-ism?)

File: vagrant/dev/bootstrap.sh [hunk 5]
```diff
@@ -12,6 +12,8 @@ yum install -y epel-release
 yum install -y https://s3.amazonaws.com/clusterhq-archive/centos/clusterhq-release$(rpm -E %dist).noarch.rpm
 
 # Install packages
+# XXX: We're disabling SELinux now so libselinux-devel may no longer be needed.
+# See https://clusterhq.atlassian.net/browse/FLOC-1643
```
Comment (exarkun) [inline vagrant/dev/bootstrap.sh]: FLOC-1643 is being resolved now.  This comment should be resolved or this reference should point somewhere else.  Does the box build if you remove that package (does the box build at all - I think it's broken in other ways at the moment ...)?

**Comment by exarkun** [comment]
Thanks.  Looks pretty good.  Buildbot has some red but there's a ton of extra builders there compared to the last time I looked at it ... so ... I think their failures can be ignored (until someone shows up and explains why new failing builders have been added ...).  Please address inline comments and merge.

**Comment by wallrj** [comment]
Hmm. The digitalocean failure might be important:

 * http://build.clusterhq.com/builders/flocker%2Facceptance%2Fdigitalocean%2Ffedora-20/builds/718/steps/trial/logs/stdio
```
[root@104.131.97.105] run: setenforce 0
[root@104.131.97.105] out: setenforce: SELinux is disabled
[root@104.131.97.105] out: 

Fatal error: run() received nonzero return code 1 while executing!

Requested: setenforce 0
Executed: /bin/bash -l -c "setenforce 0"

Aborting.
```

Looks like `setenforce 0` returns non-zero status if SELinux is already disabled...which doesn't really matter from the perspective of a user following the documentation, but it will always break our digitalocean acceptance tests.
I can do `selinuxenabled && setenforce 0` instead, but it's not as readable.

**Comment by wallrj** [comment]
Finally, all the tests have passed, apart from a few spurious Docker cleanup failures.
So I'll merge this and open another PR to merge it into master.

**Pull request merged**